stm32: addition of the STM32 JPEG Codec (NV12 -> JPEG encoding)

avolmat-st opened the pull request:
The PR adds an initial version of the STM32 JPEG Codec, available on several STM32. This PR adds it within the N6.

While the HW can perform both encoding & decoding, for the time being only encoding is provided.
The HW needs data in MCU format in order to encore into JPEG. Formating of the data into MCU is provided only for the NV12 for the time being (for that purpose probably only NV12 should be kept in this driver in order to avoid confusion).

DMA can also be used in order to inject data into the codec but this is not implemented in this version.

Main purpose of this PR is actually to provide material in order to discuss about a possible abstraction for the m2m buffers.
The driver already provide several structures which embed k_fifo but it also make it clear that m2m helper functions could be possible in order to handle in a generic way checking about in & out buffer, and have the helper call a new m2m codec api which would be in charge to perform the conversion only.
As shown in this code, the queue / dequeue function do not have anything HW specific and also take care of pushing / pulling into / from the right fifo and triggering a codec run if all condition are matched.

I hope this could generate discussion about this and based on that I can provide some helpers to hide the non HW specific code (buffer handling).

This has been tested using the tcpclientsink application modified within the PR #95862.
I however had to perform some more changes into the application, especially in order to properly set the in / out formats of the encoded video device. I provide those modifications for reference within this commit (https://github.com/avolmat-st/zephyr/commit/86249bf9b580646f3de67395d36c9ae39375ef06) part of my git repo.

This has been tested successfully on the STM32N6 connected over network with laptop into which following command is ran in order to grab JPEG data:
```
gst-launch-1.0 tcpclientsrc host=192.0.2.1 port=5000 ! jpegdec ! videoconvert ! autovideosink
```

Commit:
dts-bindings: video: addition of ST JPEG HW codec description

Add description of the ST JPEG HW codec IP node.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: stm32n6: add jpeg codec node in stm32n6.dtsi

Add the node describing the JPEG HW codec within the stm32n6

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: stm32: addition of the STM32 JPEG HW Codec support

Initial version of the support for the STM32 JPEG HW codec,
currently supporting only NV12 to JPEG without DMA support
and using SW based conversion from NV12 to MCU required
for the JPEG codec.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

mathieuchopstm commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,589 @@
+	if (data->m2m.in.fmt.pixelformat != VIDEO_PIX_FMT_JPEG) {
+		struct stm32_jpeg_fmt_conf *conf =
+			stm32_jpeg_get_conf(data->m2m.in.fmt.pixelformat);
+		HAL_StatusTypeDef hret;
```
There are some inconsistencies in where variables are declared. Either move all of them to top-of-function, or move `JPEG_ConfTypeDef Conf` declaration to here.

avolmat-st replied:
Ouch, that was an ugly one, sorry about that. Didn't notice when moving the variable.

mathieuchopstm commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,589 @@
+	struct stm32_jpeg_data *data =
+			CONTAINER_OF(hjpeg, struct stm32_jpeg_data, hjpeg);
+
+	/* Convert one MCU at a time */
+	stm32_jpeg_nv12_to_ycbcr_mcu(data->current_x_mcu++, data->current_y_mcu,
+				     data->current_in->buffer, data->current_in->buffer +
+				     data->m2m.in.fmt.width * data->m2m.in.fmt.height,
+				     data->mcu_ycbcr, data->m2m.in.fmt.width);
+	if (data->current_x_mcu >= data->m2m.in.fmt.width / MCU_WIDTH) {
+		data->current_x_mcu = 0;
+		data->current_y_mcu++;
+	}
```
Seems to be an exact duplicate from L204 - maybe it could be factored in a common function.

mathieuchopstm commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,589 @@
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	/* Check that input / output formats are correct */
+	if ((data->m2m.in.fmt.pixelformat == VIDEO_PIX_FMT_JPEG &&
+	     data->m2m.out.fmt.pixelformat == VIDEO_PIX_FMT_JPEG) ||
+	    (data->m2m.in.fmt.pixelformat != VIDEO_PIX_FMT_JPEG &&
+	     data->m2m.out.fmt.pixelformat != VIDEO_PIX_FMT_JPEG)) {
+		LOG_ERR("One of input or output format must be JPEG");
```
This is a bit difficult to follow IMO. How about:
```suggestion
        /* shorthands, not *necessary* but helps readability */
        uint32_t in_fmt = data->m2m.in.fmt.pixelformat;
        uint32_t out_fmt = data->m2m.out.fmt.pixelformat;

	if (!((in_fmt == VIDEO_PIX_FMT_JPEG && out_fmt != VIDEO_PIX_FMT_JPEG)
	     || (in_fmt != VIDEO_PIX_FMT_JPEG && out_fmt == VIDEO_PIX_FMT_JPEG))) {
		LOG_ERR("Exactly one of input/output format must be JPEG");
```

mathieuchopstm commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,589 @@
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	/* Check that input / output formats are correct */
+	if ((data->m2m.in.fmt.pixelformat == VIDEO_PIX_FMT_JPEG &&
+	     data->m2m.out.fmt.pixelformat == VIDEO_PIX_FMT_JPEG) ||
+	    (data->m2m.in.fmt.pixelformat != VIDEO_PIX_FMT_JPEG &&
+	     data->m2m.out.fmt.pixelformat != VIDEO_PIX_FMT_JPEG)) {
+		LOG_ERR("One of input or output format must be JPEG");
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* FIXME - temporary until the decoder support get added */
+	if (data->m2m.in.fmt.pixelformat == VIDEO_PIX_FMT_JPEG) {
+		LOG_ERR("Decoder not yet implemented");
+		ret = -EIO;
+		goto out;
+	}
```
Nit: These two checks operate on global data rather than `common` so maybe they can be done somewhere else.

avolmat-st replied:
Usually we do those checks at the set_stream time right before starting the device. I could have also block (the decoder) at set_fmt time, but in such case I have need to distinguish input / output in the supported format table so I found it less intrusive to put this FIXME (and the check) here, considering that this JPEG check is going to disappear when the decoder will be implemented

mathieuchopstm commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,589 @@
+
+	data->dev = dev;
+
+	/* Enable DCMIPP / CSI clocks */
```
Forgot to update comment?

mathieuchopstm commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,589 @@
+	ret = video_init_ctrl(&data->jpeg_quality, dev, VIDEO_CID_JPEG_COMPRESSION_QUALITY,
+			      (struct video_ctrl_range) {.min = 5, .max = 100,
+							 .step = 1, .def = 50});
+	if (ret) {
```
Nit: for consistency
```suggestion
	if (ret < 0) {
```

mathieuchopstm commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,589 @@
+	};									\
+										\
+	static const struct stm32_jpeg_config stm32_jpeg_config_##n = {		\
+		.jpeg_hclken =							\
+			{.bus = DT_CLOCKS_CELL(DT_DRV_INST(n), bus),		\
+			 .enr = DT_CLOCKS_CELL(DT_DRV_INST(n), bits)},		\
```
```suggestion
		.jpeg_hclken = STM32_DT_INST_CLOCKS(n), \
```

mathieuchopstm commented on snippets/stm32/video-stm32-jpegenc/README.rst:
```diff
@@ -0,0 +1,19 @@
+Overview
+********
+
+This snippet instantiate the STM32 JPEG encoder and set it
```
```suggestion
This snippet instantiate the STM32 JPEG encoder and sets it
```

mathieuchopstm commented on dts/bindings/video/st,stm32-jpeg.yaml:
```diff
@@ -0,0 +1,27 @@
+#
+
+description: |
+    STM32 JPEG Codec. The STM32 JPEG codec allows JPEG encoding and
+    decoding from / to MCU blocks which can be generated from NV12
+    format.
```
```suggestion
    STM32 JPEG Codec.
    
    The STM32 JPEG codec allows JPEG encoding and
    decoding from / to MCU blocks which can be generated from NV12
    format.
```

Maybe description can be made more generic too - from RM0486, it seems the JPEC codec is more general purpose than described here:
```suggestion
    STM32 JPEG Codec.
    
    The STM32 JPEG codec can decode JPEG-compressed images
    and encode uncompressed images to the JPEG format.
```
(Nit: `NV12` was confusing to me because RM only mentions `YUV`)

mathieuchopstm reviewed (commented):
Overall seems OK minus points raised by Etienne

Commit:
boards: stm32n6570_dk: enable the JPEG HW codec

Enable the JPEG HW codec on the stm32n6570_dk board.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: video: tcpserversink: add stm32n6570_dk jpeg conf

Add conf / overlay files in order to enable the jpeg encoder
on the stm32n6570_dk.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah reviewed (commented):
LGTM but still some questions for a small amount of minor details.
Thanks for adding support for a video encoder, good for experimenting with M2M APIs.

etienne-lms reviewed (commented):
Some minor comments.

avolmat-st commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,600 @@
+{
+	struct stm32_jpeg_data *data = dev->data;
+	struct video_common_header *common =
+		(*vbuf)->type == VIDEO_BUF_TYPE_INPUT ? &data->m2m.in : &data->m2m.out;
```
Hum, should I return -EINVAL if the type if != from input / output ? I understand your point, for sure.

avolmat-st commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,600 @@
+		.height_min = 1,
+		.height_max = 65535,
+		.width_step = 1,
+		.height_step = 1,
```
Yes, corrected

avolmat-st commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,600 @@
+	/* FIXME - temporary until the decoder support get added */
+	if (data->m2m.in.fmt.pixelformat == VIDEO_PIX_FMT_JPEG) {
+		LOG_ERR("Decoder not yet implemented");
+		ret = -EIO;
```
I removed this code following other review comment above

avolmat-st commented:
I also added a new commit at the beginning to change the input/output setting order in the encoder. Indeed, I feel it make sense to ensure that the input is set PRIOR to setting the output since the driver might adapt the output settings after having received the input settings. Does that make sense ?

(VENC still work with this modification, and this also already the way I coded it into the UVC encoder sample)

erwango commented on drivers/video/Kconfig.stm32_jpeg:
```diff
@@ -0,0 +1,17 @@
+	select RESET
+	select USE_STM32_HAL_JPEG
+	select USE_STM32_HAL_DMA
+	select USE_STM32_HAL_DMA_EX
```
Not required then, I guess

avolmat-st replied:
Actually it is needed. The JPEG HAL code embeds calls to HAL DMA functions so it is necessary in order to properly build.

ngphibang commented on drivers/video/video_stm32_jpeg.c:
```diff
@@ -0,0 +1,600 @@
+
+		/* JPEG Encoding */
+		jpeg_conf.ColorSpace = JPEG_YCBCR_COLORSPACE;
+		jpeg_conf.ChromaSubsampling = conf->subsampling;
```
Should we need to limit the use of `__ASSERT_NO_MSG() ` as it will / should not go into production code (should use it only if we are in a void function) ?

avolmat-st commented:
> > I also added a new commit at the beginning to change the input/output setting order in the encoder. Indeed, I feel it make sense to ensure that the input is set PRIOR to setting the output since the driver might adapt the output settings after having received the input settings. Does that make sense ?
> 
> I think it can be done in either ways so the commit is OK if we see it is more convenient that way in the sample. BTW, I think the drivers should memorize and check the input and output formats so that they are compatible despite of the order. The important thing is we should not enforce the order in the application.
> 
> (For example, in libMP or gstreamer, caps negotiation can be fixed between the output and the downstream first then between the input and the upstream, or vice versa. Also, during caps negotiation, to transform a caps from one side to another, we also call set_format() on both sides to check if an input/output format can go with the other one)

Hum, in such case I will have to rework the way done now in the last version of the JPEG driver I pushed. If we don't mandate to have first the input, then the output, at set_fmt time few type checking can be done and this needs to be moved to the set_stream time (a bit as I did in the first version).
Indeed, when setting one side it is not possible to check the other side since the other side might still have to be set. For example, assume that by default the driver is in encoder mode, aka NV12 -> JPEG, when changing it to become a decoder, then we might set the output to NV12, which leads to an invalid config of NV12 -> NV12. Same for the input as well, since JPEG -> JPEG is invalid. So in such case overall configuration cannot be checked at set_fmt time and this has to be deferred.

The merit (I think) of forcing the input first is that, whenever setting first the input, the driver will update the output to a valid configuration. Then the application can check if it is ok for the use-case and change it if not. Also set_fmt time check can be done.

avolmat-st commented:
PR simply rebased on top of HEAD to fix merge conflict (due to conflict in the stm32n6_dk_common.dtsi)

josuah reviewed (approved):
Nothing blocking left I could think of for my side.

It will be interesting to see how format negotiation can happen in the future for M2M and other long pipeline of formats.
